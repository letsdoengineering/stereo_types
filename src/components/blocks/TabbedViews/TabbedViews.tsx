import React, { useState } from 'react'
import TabButton from './TabButton/TabButton'
import GridColumn from '../../basics/Grid/GridColumn/GridColumn'
import GridRow from '../../basics/Grid/GridRow/GridRow'
import tabbedViewsStyles from './TabbedViewsStyles.module.css'
import Spacing from '../../basics/Spacing/Spacing'

type TabbedViewsProps = {
  /** string description/view element/component key/pairs - worth noting this is specifically
   * an object and not an array of arrays because the assumption is we'd like to ensure unique tab names*/
  views: Record<string, JSX.Element | React.ReactElement>
}

/** TabbedViews: takes a 'views' object consisting of key/pairs of the tab
 * description and the corresponding view and returns a tabbed interface for
 * selecting different views */
const TabbedViews: React.FC<TabbedViewsProps> = ({ views }: TabbedViewsProps) => {
  const [[currentViewName, currentView], setCurrentView] = useState<
    [string, JSX.Element | React.ReactElement]
  >(Object.entries(views)[0])

  const tabButtons = Object.entries(views).map((el, index) => {
    const [viewName, view] = el
    const isActive = view === currentView
    const viewNameForButtonAria = viewName.replace(/\s/g, '')
    return (
      <GridColumn role='tablist' size='1' key={`${viewName}-${index}-tab-button-container`}>
        <TabButton
          key={`${viewName}-${index}-tab-button`}
          onClick={(): void => setCurrentView([viewName, view])}
          isHighlighted={isActive}
          role='tab'
          aria-selected={isActive}
          aria-controls={`${viewNameForButtonAria}-tab`}
          id={`${viewNameForButtonAria}-button`}
        >
          {viewName}
        </TabButton>
      </GridColumn>
    )
  })

  const viewNameForTabAria = currentViewName.replace(/\s/g, '')
  return (
    <>
      <GridRow className={tabbedViewsStyles.tabSection}>{tabButtons}</GridRow>
      <Spacing size='Double' />
      <div id={`${viewNameForTabAria}-tab`} aria-labelledby={`${viewNameForTabAria}-button`}>
        {currentView}
      </div>
    </>
  )
}

export default TabbedViews
