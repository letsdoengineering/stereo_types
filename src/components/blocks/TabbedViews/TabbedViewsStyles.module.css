.tab-section {
    border-bottom: 1px solid var(--outline-gray);
}
