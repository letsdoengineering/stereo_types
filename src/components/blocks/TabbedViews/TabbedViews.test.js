import * as React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'
import TabbedViews from './TabbedViews'

const testViews = {
    'view 1': <div>some content for default view</div>,
    'view 2': <div>some content for secondary view</div>,
}

describe('TabbedViews', () => {
    it('should present the user with tabbable views', () => {
        render(<TabbedViews views={testViews} />)

        const tab1 = screen.getByText('view 1')
        const tab2 = screen.getByText('view 2')

        // initial render - should default to view 1, with no others available
        expect(screen.queryByText('some content for default view')).toBeInTheDocument()
        expect(screen.queryByText('some content for secondary view')).not.toBeInTheDocument()

        // click tab for view 2 - should display view two and no others
        userEvent.click(tab2)

        expect(screen.queryByText('some content for secondary view')).toBeInTheDocument()
        expect(screen.queryByText('some content for default view')).not.toBeInTheDocument()

        // click tab for view 1 to return to that view - should not display any other view
        userEvent.click(tab1)

        expect(screen.queryByText('some content for default view')).toBeInTheDocument()
        expect(screen.queryByText('some content for secondary view')).not.toBeInTheDocument()
    })

    it('has not changed the JSX layout of the page from previous snapshot', () => {
        const tree = renderer.create(<TabbedViews views={testViews} />).toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const { container } = render(<TabbedViews views={testViews} />)
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
