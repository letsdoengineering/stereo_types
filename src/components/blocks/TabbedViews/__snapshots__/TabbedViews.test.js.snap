// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabbedViews has not changed the JSX layout of the page from previous snapshot 1`] = `
Array [
  <div
    className="gridRow tabSection"
  >
    <div
      className="gridColumn size1"
      role="tablist"
    >
      <button
        aria-controls="view1-tab"
        aria-selected={true}
        className="tabButton highlighted"
        id="view1-button"
        onClick={[Function]}
        role="tab"
        type="button"
      >
        <span
          className="sizeS weightNormal textGray"
        >
          view 1
        </span>
      </button>
    </div>
    <div
      className="gridColumn size1"
      role="tablist"
    >
      <button
        aria-controls="view2-tab"
        aria-selected={false}
        className="tabButton"
        id="view2-button"
        onClick={[Function]}
        role="tab"
        type="button"
      >
        <span
          className="sizeS weightNormal textGray"
        >
          view 2
        </span>
      </button>
    </div>
  </div>,
  <div
    className="spacingDouble spacing"
  />,
  <div
    aria-labelledby="view1-button"
    id="view1-tab"
  >
    <div>
      some content for default view
    </div>
  </div>,
]
`;
